Fenrir/Commands.fs

```diff
@@ -159,3 +165,86 @@ let stringToByte (s: String): byte[] =
+    Array.iter printAtom tree
+
+let changeHashInCommit (commit: CommitBody) (hash: byte[]): CommitBody =
+    {Tree = hash |> byteToString;
+     Parents = commit.Parents;
+     Rest = commit.Rest}
+
+let commitBodyToStream (commit: CommitBody) (stream: Stream): unit =
+    let printParent (a: String): unit =
+        stream.Write(ReadOnlySpan<byte>("parent "B))
+        stream.Write(ReadOnlySpan<byte>(a |> Encoding.ASCII.GetBytes))
+        stream.WriteByte('\n'B)
+
+    stream.Write(ReadOnlySpan<byte>("tree "B))
+    stream.Write(ReadOnlySpan<byte>(commit.Tree |> Encoding.ASCII.GetBytes))
+    stream.WriteByte('\n'B)
+
+    Array.iter printParent commit.Parents
+    stream.Write(ReadOnlySpan<byte>(String.Join('\n', commit.Rest) |> Encoding.ASCII.GetBytes))
+
+type TreeStreams (length: int) =
+    member this.Streams = Array.create (length) (new MemoryStream())
+    member this.Hashes = Array.create (length) ""
+    interface System.IDisposable with
+        member this.Dispose() =
+            Array.iter (fun (s: MemoryStream) -> s.Dispose()) this.Streams
+
+let updateObjectInTree (rootTreeHash: string) (pathToRepo: String) (filePath: string) (blobHash: string): TreeStreams =
+    let filePathList = filePath.Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries) |> List.ofArray
+    use treeStreams = new TreeStreams (filePathList.Length)
+
+    let rec updateFileHashInTree (tree: TreeBody) (filePaths: String list): String =
+        let index = treeStreams.Streams.Length - filePaths.Length
+        match filePaths with
+        | [] -> failwithf "Empty path to file"
+        | [fileName] ->
+            let newTree = changeHashInTree tree (stringToByte blobHash) fileName
+
+            use input = new MemoryStream()
+            treeBodyToStream newTree input
+            writeObjectHeader GitObjectType.GitTree input treeStreams.Streams.[index]
+            let hash = SHA1 treeStreams.Streams.[index] |> byteToString
+            treeStreams.Streams.[index].Position <- 0L
+            treeStreams.Hashes.[index] <- hash
+            hash
+        | directoryName :: restPathSegments ->
+            let directoryHash = hashOfObjectInTree tree directoryName |> byteToString
+            let subTree = parseTreeBody pathToRepo directoryHash
+            let newHash = updateFileHashInTree subTree restPathSegments
+            let newTree = changeHashInTree tree (stringToByte newHash) directoryName
+
+            use input = new MemoryStream()
+            treeBodyToStream newTree input
+            writeObjectHeader GitObjectType.GitTree input treeStreams.Streams.[index]
+            let hash = SHA1 treeStreams.Streams.[index] |> byteToString
+            treeStreams.Streams.[index].Position <- 0L
+            treeStreams.Hashes.[index] <- hash
+            hash
+
+    let parentTree = parseTreeBody pathToRepo rootTreeHash
+    updateFileHashInTree parentTree filePathList |> ignore
+    treeStreams
```

Comment:
Maybe it's a good idea to reset all the streams inside of `treeStreams` to make them readable (something like `treeStreams.Streams |> Array.iter (fun s -> s.Position <- 0L)`).